### ISSUE TITLE
Inconsistent behavior when querying a relatively high number of rows

### ISSUE THREAD

**Issue by lcorbel** [snowflakedb/snowflake-connector-python#255]
Hi,

I would like to report an issue I have been experiencing. It seems that the connector has an inconsistent behavior when querying a relatively big number of rows (~10 000 rows) in a select statement. Sometimes the query will go through and return the result and sometimes I will get an error message instead that you can see described below. I have a feeling this is linked to some kind of hard limit or delay to respect between queries but I cannot find any mention of it in the documentation nor can I explain the inconsistent behavior. If I had a 'LIMIT 100' statement at the end of my query it will succeed every time.

Many thanks in advance!

Please answer these questions before submitting your issue. Thanks!

What version of Python are you using (python --version)?
python3 --version Python 3.5.2
What operating system and processor architecture are you using (python -c 'import platform; print(platform.platform())')?
python -c 'import platform; print(platform.platform())' Linux-4.8.0-53-generic-x86_64-with-LinuxMint-18.2-sonya
What are the component versions in the environment (pip list)?
asn1crypto 1.2.0 azure-common 1.1.23 azure-storage-blob 2.1.0 azure-storage-common 2.1.0 boto3 1.9.253 botocore 1.12.253 certifi 2019.9.11 cffi 1.13.2 cryptography 2.8 idna 2.0 ijson 2.5.1 oscrypto 1.1.0 pip 19.3.1 pycrypto 2.6.1 pycryptodomex 3.9.3 pycurl 7.43.0 PyJWT 1.3.0 pyOpenSSL 19.0.0 pytz 2019.3 requests 2.9.1 six 1.10.0 snowflake-connector-python 2.1.2 urllib3 1.24.3
What did you do?
If possible, provide a recipe for reproducing the error.
A complete runnable program is good.
`
import snowflake.connector
import pandas as pd
from google.cloud import storage

ctx = snowflake.connector.connect(
    user='<user>',
    password='<pass>',
    account='<acc>'
    )
cs = ctx.cursor()
try:
    cs.execute("USE WAREHOUSE <warehouse>")
    cs.execute("USE <db>")
    cs.execute("USE SCHEMA <schema>")

    cols = ['IDENTIFIANT', 'CANAL', 'STATUT_EXECUTION', 'LIBELLE']
    
    query = "SELECT IDENTIFIANT, CANAL, STATUT_EXECUTION, LIBELLE FROM <table>"

    cs.execute(query)
    results = cs.fetchall()
    results_df = pd.DataFrame(results, columns=cols)

finally:
    cs.close()
ctx.close()

results_df.to_csv('results.csv')
`
What did you expect to see?
I did not expect to see any errors. Sometimes the query works perfectly fine and i get my 12444 rows output and sometimes I get the following error message.
What did you see instead?

Traceback (most recent call last):
  File "arrow_result.pyx", line 118, in snowflake.connector.arrow_result.ArrowResult.__next__
  File "arrow_iterator.pyx", line 130, in snowflake.connector.arrow_iterator.EmptyPyArrowIterator.__next__
StopIteration

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "analyses_tp.py", line 23, in <module>
    results = cs.fetchall()
  File "/home/louico/.local/lib/python3.5/site-packages/snowflake/connector/cursor.py", line 861, in fetchall
    row = self.fetchone()
  File "/home/louico/.local/lib/python3.5/site-packages/snowflake/connector/cursor.py", line 825, in fetchone
    return next(self._result)
  File "arrow_result.pyx", line 124, in snowflake.connector.arrow_result.ArrowResult.__next__
  File "/home/louico/.local/lib/python3.5/site-packages/snowflake/connector/chunk_downloader.py", line 187, in next_chunk
    raise self._downloader_error
  File "/home/louico/.local/lib/python3.5/site-packages/snowflake/connector/chunk_downloader.py", line 125, in _download_chunk
    result_data = self._fetch_chunk(self._chunks[idx].url, headers)
  File "/home/louico/.local/lib/python3.5/site-packages/snowflake/connector/chunk_downloader.py", line 258, in _fetch_chunk
    binary_data_handler=handler)
  File "/home/louico/.local/lib/python3.5/site-packages/snowflake/connector/network.py", line 612, in fetch
    **kwargs)
  File "/home/louico/.local/lib/python3.5/site-packages/snowflake/connector/network.py", line 714, in _request_exec_wrapper
    raise e
  File "/home/louico/.local/lib/python3.5/site-packages/snowflake/connector/network.py", line 656, in _request_exec_wrapper
    method, full_url, headers, data, conn)
  File "/home/louico/.local/lib/python3.5/site-packages/snowflake/connector/network.py", line 761, in _handle_unknown_error
    u'errno': ER_FAILED_TO_REQUEST,
  File "/home/louico/.local/lib/python3.5/site-packages/snowflake/connector/errors.py", line 100, in errorhandler_wrapper
    connection.errorhandler(connection, cursor, errorclass, errorvalue)
  File "/home/louico/.local/lib/python3.5/site-packages/snowflake/connector/errors.py", line 73, in default_errorhandler
    done_format_msg=errorvalue.get(u'done_format_msg'))
snowflake.connector.errors.OperationalError: 250003: Failed to get the response. Hanging? method: get, url: https://rqpukksfcb1stg.blob.core.windows.net/results/019162e9-00e9-6ad3-0000-0ced0007226a_0/main/data_0_0_0?sig=%2F1dgO5iB70R4OT8Ye7Ds0ZgR7oJgLQ%2BlzlOhGHD4atE%3D&se=2020-01-07T13%3A15%3A28Z&sv=2018-11-09&spr=https&rsce=gzip&sp=rl&sr=b

Can you set logging to DEBUG and collect the logs?
import logging
import os

for logger_name in ['snowflake.sqlalchemy', 'snowflake.connector', 'botocore']: 
    logger = logging.getLogger(logger_name)
    logger.setLevel(logging.DEBUG)
    ch = logging.StreamHandler()
    ch.setLevel(logging.DEBUG)
    ch.setFormatter(logging.Formatter('%(asctime)s - %(threadName)s %(filename)s:%(lineno)d - %(funcName)s() - %(levelname)s - %(message)s'))
    logger.addHandler(ch)

**Comment by snowstakeda** [comment]
The result set is downloaded from the Azure Blob storage if the size is more than 100k bytes. Can you ensure https://rqpukksfcb1stg.blob.core.windows.net/ is accessible by the client?

I would suggest to reach out the support in case where we need to look into the full logs.

**Comment by lcorbel** [comment]
Hey thanks for your answer. FYI when I open the URL https://rqpukksfcb1stg.blob.core.windows.net/results/019173dc-006e-19f2-0000-0ced0007404a_0/main/data_0_0_0?sig=Ej7qq%2FKVAToTV0hYX8eEcoKaMrjE%2BlXIBXRXdDtmnjI%3D&se=2020-01-10T15%3A36%3A30Z&sv=2018-11-09&spr=https&rsce=gzip&sp=rl&sr=b in my browser I get an XMl file conrtaining the following error message : "The blob is encrypted with customer specified encryption, but it was not provided in the request. RequestId:<id of the request>".

But what is really bugging me is that sometimes I don't get an error message at all I get the result of my query as expected so it means the Azure Blob storage is not inaccessible from the client.

I would love to see the full logs but I cannot open a support ticket as I am not registered as a Snowflake Lodge user.

**Comment by snowstakeda** [comment]
Ok, that's the expected error message when you open from the browser. I still think there are some connectivity issue from the python connector application which don't have enough access to the required resources.

Could you follow the steps:
- Run `snowcd` from the client where the python application runs: https://docs.snowflake.net/manuals/user-guide/snowcd.html.
- Run curl command. HTTP 400 is the expected result because no parameter is passed.
```
curl -v https://rqpukksfcb1stg.blob.core.windows.net/
```
- Run openssl command:
```
echo QUIT | openssl s_client -connect rqpukksfcb1stg.blob.core.windows.net:443 -showcerts > certs.txt
```
The output should be:
```
depth=2 C = IE, O = Baltimore, OU = CyberTrust, CN = Baltimore CyberTrust Root
verify return:1
depth=1 C = US, ST = Washington, L = Redmond, O = Microsoft Corporation, OU = Microsoft IT, CN = Microsoft IT TLS CA 2
verify return:1
depth=0 CN = *.blob.core.windows.net
verify return:1
DONE
```
If any of the above didn't work, there must be some proxy/firewall issue in your environment.

**Comment by lcorbel** [comment]
Hi,

I ran the command you described and I am getting every expected output. So it does not seem to be a proxy/firewall issue.

I did some more experimenting and I noticed that the first time I run my query in a while, it always succeed and return the expected results but then if I try to run the same query directly after I get the "snowflake.connector.errors.OperationalError: 250003: Failed to get the response" error. So it looks like there is some kind threshold I am going over by requesting too much data from the blob storage in too small a window of time.

Many thanks again for helping me out.
